Address issue #1139: make --force-reinstall not require --upgrade

cjerdonek opened the pull request:
Fixes #1139.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

pradyunsg commented:
@cjerdonek Could you rebase this PR on the current master?

cjerdonek commented:
@pradyunsg Sure, thanks for noticing this.

By the way, I noticed some time ago that it looked like you were doing some significant refactoring. Is that ongoing or is it at a natural stopping point?

cjerdonek commented:
@pradyunsg Okay, this is rebased now. A few comments on the patch commits:

1. The first several commits don't change the functionality at all. They're just simple preliminary refactorings that I kept separate for your benefit. I confirmed on Travis that the tests passed after these initial preparatory commits.
2. In the second-last commit, I added the failing tests for the issue. I confirmed on Travis that these new tests fail on this commit.
3. In the last commit, I made the behavior change re: `--force-install` and confirmed that all the tests (including the new ones) pass on Travis.

cjerdonek commented:
Also, [here](https://github.com/cjerdonek/pip/commits/issue_1139_force_reinstall_v2) is where you can see the Travis CI results I mentioned above for those prior commits.

cjerdonek commented:
Okay, I will create a separate PR then. Note that I wouldn't call it a "style change" PR though as it does things like create a separate function called `_set_req_to_reinstall()` to make the existing code more DRY (a function which is useful for the later functional change). It also simplifies the code and alters the if structure to make the later functional change easier to reason about.

pradyunsg commented:
That I would consider a functional change... The change in skip_reason generation is a style-change. :)

cjerdonek commented:
For the benefit of readers of this thread, I created PR #4644 as @pradyunsg requested. He approved it, so once it's merged, I'll rebase this PR.

cjerdonek commented:
Rebased now that PR #4644 is merged.

cjerdonek commented:
@pradyunsg I noticed that the "pypy" Travis CI tests have consistently been failing for me on my PR's for reasons unrelated to the code in my PR (and occasionally for other environments in the Travis CI runs), even if I retrigger the tests several times with an empty `git commit --amend`. Will PR's not be considered if they have a red X next to them, even if the X is due to test flakiness?

pradyunsg commented:
> I noticed that the "pypy" Travis CI tests have consistently been failing for me on my PR's for reasons unrelated to the code in my PR

It's happening with almost all builds. It seems to stem from some issue in pytest-xdist that causes it to hang subprocesses. I've been trying to figure out how to fix/workaround that for the past few days... 

> Will PR's not be considered if they have a red X next to them, even if the X is due to test flakiness?

TBH, I don't really know about this. Maybe @dstufft or @pfmoore would have something to say about that.

pradyunsg commented on tests/functional/test_install_force_reinstall.py:
```diff
@@ -0,0 +1,45 @@
+
+def check_installed_version(script, package, expected):
+    # Python 3.3 emits a "Python 3.3 support has been deprecated" warning.
+    expect_stderr = (pyversion_tuple[:2] == (3, 3))
```
IIRC, this is already handled somewhere within script.

pradyunsg commented on tests/functional/test_install_force_reinstall.py:
```diff
@@ -0,0 +1,45 @@
+def check_installed_version(script, package, expected):
+    # Python 3.3 emits a "Python 3.3 support has been deprecated" warning.
+    expect_stderr = (pyversion_tuple[:2] == (3, 3))
+    result = script.run('pip', 'show', package, expect_stderr=expect_stderr)
```
`script.pip` would be nicer?

pradyunsg commented on tests/functional/test_install_force_reinstall.py:
```diff
@@ -0,0 +1,45 @@
+    for line in lines:
+        if line.startswith('Version: '):
+            version = line.split()[-1]
+            break
```
This should probably fail if it doesn't find version with a nice message instead of a NameError...

pradyunsg commented on tests/functional/test_install_force_reinstall.py:
```diff
@@ -0,0 +1,45 @@
+    """
+    result = script.pip('install', 'INITools==0.2')
+    check_installed_version(script, 'initools', '0.2')
+    result2 = script.pip('install', '--force-reinstall', 'INITools')
```
style: If you could add empty lines above the not-first script.pip calls, that would be nice, since it kinda visually divides the runs.

pradyunsg commented:
> It's happening with almost all builds.

FWIW, it seems to have settled down and fixed itself? I have bumped into it for my last few builds.

_hisenbug_

cjerdonek commented:
Thanks for reviewing, @pradyunsg! I updated the PR and addressed your comments (and also made the two tests DRY).

cjerdonek commented:
> FWIW, it seems to have settled down and fixed itself?

It seems still to be happening, FYI, like with my update to this PR just now.

cjerdonek commented:
@pradyunsg Can you also take a look at this now that I've addressed your comments? Thanks.

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

cjerdonek commented:
Rebased.

xavfernandez commented on tests/functional/test_install_force_reinstall.py:
```diff
@@ -0,0 +1,49 @@
+    check_force_reinstall(script, 'INITools', '0.3.1')
+
+
+@pytest.mark.network
```
Instead of using `INITools` which needs a network access, you could test it with `simplewheel` in https://github.com/pypa/pip/tree/master/tests/data/packages and `--find-link`/`--no-index`.

cjerdonek replied:
Thanks. Can you point to an example in the code of the preferred pattern we should be using? I used `INITools` after looking at similar tests. It would be good to make it easier for contributors to write faster tests.

pradyunsg replied:
Have a look here?

https://github.com/pypa/pip/blob/d339351a41846315249ff8edaf9e8569d03f6d8c/tests/functional/test_install.py#L202

xavfernandez replied:
Indeed 👍 
cf https://github.com/pypa/pip/blob/d339351a41846315249ff8edaf9e8569d03f6d8c/tests/lib/__init__.py#L364-L369 for definition

cjerdonek commented:
Updated.

cjerdonek commented:
@pradyunsg @xavfernandez With pip 10 nearing, would either of you mind reviewing this patch once more? It addresses a 4-year old bug. I addressed all review comments from both of you earlier. Thanks if you can!

pradyunsg commented:
Thanks for the reminder. I'll try to do it in the evening today.

cjerdonek commented:
Thanks. The only spot I found with a search is [this mention](https://github.com/pypa/pip/blame/37f236de70fd22782c2948362ff0ed0ba2cf1d17/docs/user_guide.rst#L681) in the User Guide, which isn't primarily about `--force-reinstall`:

    You can then install from the archive like this::

        $ tempdir=$(mktemp -d /tmp/wheelhouse-XXXXX)
        $ (cd $tempdir; tar -xvf /path/to/bundled.tar.bz2)
        $ pip install --force-reinstall --ignore-installed --upgrade --no-index --no-deps $tempdir/*

The `--upgrade` probably becomes superfluous with this PR (but it doesn't hurt to keep it in).

pradyunsg commented:
I'll wait while (a couple of weeks?) for @xavfernandez to review it. :)

BrownTruck commented:
Hello!

I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the ``master`` branch into this pull request or rebase this pull request against ``master`` then it will eligible for code review and hopefully merging!

xavfernandez commented:
Thanks again :)

cjerdonek commented:
Thanks!
